mlflow/sklearn/__init__.py

```diff
@@ -1494,10 +1496,24 @@ def should_patch_estimator_method(class_def, func_name):
                 FLAVOR_NAME, class_def, "score", patched_model_score, manage_run=False,
             )
 
-    from sklearn import metrics
-
     for metric_name in _get_metric_name_list():
-        safe_patch(FLAVOR_NAME, metrics, metric_name, patched_metric_api, manage_run=False)
+        safe_patch(FLAVOR_NAME, sklearn.metrics, metric_name, patched_metric_api, manage_run=False)
+
+    def scorer_finalizer(scorer_, original_score_fn_):
+        scorer_._score_func = original_score_fn_
+
+    for scorer in sklearn.metrics.SCORERS.values():
+        scorer_fn = scorer._score_func
+        scorer_fn_name = scorer_fn.__name__
+        try:
+            original_metric_fn = gorilla.get_original_attribute(sklearn.metrics, scorer_fn_name)
+        except AttributeError:
+            original_metric_fn = None
+        if scorer_fn is original_metric_fn:
+            scorer._score_func = getattr(sklearn.metrics, scorer_fn_name)
+            gorilla.get_active_patch(sklearn.metrics, scorer_fn_name).add_finalizer(
+                scorer_finalizer, scorer, original_metric_fn
+            )
```

Comment:
Instead of adding finalizers, could we iterate through each of the scorers and `safe_patch` its `score_func` attribute?

```
for scorer in sklearn.metrics.SCORERS.values():
    safe_patch(FLAVOR_NAME, scorer, "_score_func", patched_metric_api, manage_run=False)
```

This seems to pass the tests you included in the PR.

Reply:
This is a good idea! Then no need adding hook when reverting patch.